ledger/simulation/simulator.go

```diff
@@ -0,0 +1,168 @@
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package simulation
+
+import (
+	"errors"
+
+	"github.com/algorand/go-algorand/data"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/verify"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+)
+
+// ==============================
+// > Simulator Ledger
+// ==============================
+
+// simulatorLedger patches the ledger interface to use a constant latest round.
+type simulatorLedger struct {
+	*data.Ledger
+	start basics.Round
+}
+
+// Latest is part of the LedgerForSimulator interface.
+// We override this to use the set latest to prevent racing with the network
+func (l simulatorLedger) Latest() basics.Round {
+	return l.start
+}
+
+// LookupLatest would implicitly use the latest round in the _underlying_
+// Ledger, it would give wrong results if that ledger has moved forward. But it
+// should never be called, as the REST API is the only code using this function,
+// and the REST API should never have access to a simulatorLedger.
+func (l simulatorLedger) LookupLatest(addr basics.Address) (basics.AccountData, basics.Round, basics.MicroAlgos, error) {
+	err := errors.New("unexpected call to LookupLatest")
+	return basics.AccountData{}, 0, basics.MicroAlgos{}, err
+}
+
+// ==============================
+// > Simulator Errors
+// ==============================
+
+// SimulatorError is the base error type for all simulator errors.
+type SimulatorError struct {
+	err error
+}
+
+func (s SimulatorError) Error() string {
+	return s.err.Error()
+}
+
+func (s SimulatorError) Unwrap() error {
+	return s.err
+}
+
+// InvalidTxGroupError occurs when an invalid transaction group was submitted to the simulator.
+type InvalidTxGroupError struct {
+	SimulatorError
+}
+
+// EvalFailureError represents an error that occurred during evaluation.
+type EvalFailureError struct {
+	SimulatorError
+}
+
+// ScopedSimulatorError is a simulator error that has 2 errors, one for internal use and one for
+// displaying publicly. THe external error is useful for API routes/etc.
+type ScopedSimulatorError struct {
+	SimulatorError        // the original error for internal use
+	External       string // the external error for public use
+}
```

Comment:
I recall using this `ScopedSimulatorError` somewhere but it could have been removed since then. I'll take a look and remove this if that's not the case. This is likely an artifact of the previous version of the implementation which was used directly by the API handler.

Reply:
You are correct, it appears to no longer be used. I'll remove it.